core/filterconstraint.cpp

```diff
@@ -0,0 +1,939 @@
+};
+
+static struct string_mode_description {
+	filter_constraint_string_mode mode;
+	const char *s;
+	const char *translated;
+} string_mode_descriptions[] = {
+	{ FILTER_CONSTRAINT_STARTS_WITH, "starts_with", QT_TRANSLATE_NOOP("gettextFromC", "starting with") },
+	{ FILTER_CONSTRAINT_SUBSTRING, "substring", QT_TRANSLATE_NOOP("gettextFromC", "with substring") },
+	{ FILTER_CONSTRAINT_EXACT, "exact", QT_TRANSLATE_NOOP("gettextFromC", "exactly") },
+};
+
+static struct range_mode_description {
+	filter_constraint_range_mode mode;
+	const char *s;
+	const char *translated;
+	const char *translated_date;
+} range_mode_descriptions[] = {
+	{ FILTER_CONSTRAINT_EQUAL, "equal", QT_TRANSLATE_NOOP("gettextFromC", "equal to"), QT_TRANSLATE_NOOP("gettextFromC", "at") },
+	{ FILTER_CONSTRAINT_LESS, "less", QT_TRANSLATE_NOOP("gettextFromC", "at most"), QT_TRANSLATE_NOOP("gettextFromC", "before") },
+	{ FILTER_CONSTRAINT_GREATER, "greater", QT_TRANSLATE_NOOP("gettextFromC", "at least"), QT_TRANSLATE_NOOP("gettextFromC", "after") },
+	{ FILTER_CONSTRAINT_RANGE, "range", QT_TRANSLATE_NOOP("gettextFromC", "in range"), QT_TRANSLATE_NOOP("gettextFromC", "in range") },
+};
+
+static const char *negate_description[2] {
+	QT_TRANSLATE_NOOP("gettextFromC", "is"),
+	QT_TRANSLATE_NOOP("gettextFromC", "is not"),
+};
+
+// std::size() is only available starting in C++17, so let's roll our own.
+template <typename T, size_t N>
+static constexpr size_t array_size(const T (&)[N])
+{
+	return N;
+}
+
+static constexpr size_t type_descriptions_count()
+{
+	return array_size(type_descriptions);
+}
+
+static constexpr size_t string_mode_descriptions_count()
+{
+	return array_size(string_mode_descriptions);
+}
+
+static constexpr size_t range_mode_descriptions_count()
+{
+	return array_size(range_mode_descriptions);
+}
+
+static const type_description *get_type_description(enum filter_constraint_type type)
+{
+	for (size_t i = 0; i < type_descriptions_count(); ++i) {
+		if (type_descriptions[i].type == type)
+			return &type_descriptions[i];
+	}
+	report_error("unknown filter constraint type: %d", type);
+	return nullptr;
+}
+
+static const string_mode_description *get_string_mode_description(enum filter_constraint_string_mode mode)
+{
+	for (size_t i = 0; i < string_mode_descriptions_count(); ++i) {
+		if (string_mode_descriptions[i].mode == mode)
+			return &string_mode_descriptions[i];
+	}
+	report_error("unknown filter constraint string mode: %d", mode);
+	return nullptr;
+}
+
+static const range_mode_description *get_range_mode_description(enum filter_constraint_range_mode mode)
+{
+	for (size_t i = 0; i < range_mode_descriptions_count(); ++i) {
+		if (range_mode_descriptions[i].mode == mode)
+			return &range_mode_descriptions[i];
+	}
+	report_error("unknown filter constraint string mode: %d", mode);
+	return nullptr;
+}
+
+extern "C"
+const char *filter_constraint_type_to_string(enum filter_constraint_type type)
```

Comment:
I'm too lazy to make sure, but I think when we did single line `extern "C"` markings, we had them on the same line

Reply:
done